File: heapsort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

void heapify(int *arr, int n, int i) {
    int largest = i;         
    int left = 2 * i + 1;    
    int right = 2 * i + 2;   

    if (left < n && arr[left] > arr[largest]) {
        largest = left;
    }

    if (right < n && arr[right] > arr[largest]) {
        largest = right;
    }

    if (largest != i) {
        int temp = arr[i];
        arr[i] = arr[largest];
        arr[largest] = temp;

        heapify(arr, n, largest);
    }
}

void heap_sort(int *arr, int n) {
    for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }

    for (int i = n - 1; i > 0; i--) {
        int temp = arr[0];
        arr[0] = arr[i];
        arr[i] = temp;

        heapify(arr, i, 0);
    }
}

double mean(const double *values, int n) {
    double sum = 0.0;
    for (int i = 0; i < n; i++) {
        sum += values[i];
    }
    return sum / (double)n;
}

double stddev_sample(const double *values, int n) {
    if (n < 2) {
        return 0.0;
    }

    double m = mean(values, n);
    double sum_sq = 0.0;
    for (int i = 0; i < n; i++) {
        double diff = values[i] - m;
        sum_sq += diff * diff;
    }
    return sqrt(sum_sq / (double)(n - 1));
}


double run_single_sort(int n) {
    int *arr = (int *)malloc(n * sizeof(int));
    if (arr == NULL) {
        fprintf(stderr, "Erro ao alocar memória para n = %d\n", n);
        exit(EXIT_FAILURE);
    }

    for (int i = 0; i < n; i++) {
        arr[i] = rand();  
    }

    clock_t start = clock();
    heap_sort(arr, n);
    clock_t end = clock();

    free(arr);

    double elapsed = (double)(end - start) / (double)CLOCKS_PER_SEC;
    return elapsed;
}

int main(void) {
    srand((unsigned int)time(NULL));

    const int sizes[] = {100, 500, 1000, 5000, 10000, 50000};
    const int NUM_SIZES = sizeof(sizes) / sizeof(sizes[0]);

    const int NUM_RUNS = 20;

    printf("Experimento Heap Sort\n");
    printf("Numero de execucoes por tamanho: %d\n\n", NUM_RUNS);
    printf("n\tmedia (s)\tdesvio-padrao (s)\n");
    printf("-------------------------------------------------\n");

    for (int s = 0; s < NUM_SIZES; s++) {
        int n = sizes[s];
        double *times = (double *)malloc(NUM_RUNS * sizeof(double));
        if (times == NULL) {
            fprintf(stderr, "Erro ao alocar memoria para tempos (n = %d)\n", n);
            exit(EXIT_FAILURE);
        }

        for (int i = 0; i < NUM_RUNS; i++) {
            times[i] = run_single_sort(n);
        }

        double m = mean(times, NUM_RUNS);
        double sd = stddev_sample(times, NUM_RUNS);

        printf("%d\t%.6f\t%.6f\n", n, m, sd);

        free(times);
    }

    return 0;
}

File: heapsort.py
import random
import time
import statistics


def heapify(arr, n, i):
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    if left < n and arr[left] > arr[largest]:
        largest = left

    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  
        heapify(arr, n, largest)  
def heap_sort(arr):
    n = len(arr)

    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heapify(arr, i, 0)


def run_experiment(sizes, runs_per_size=20):
    resultados = []  

    for n in sizes:
        tempos = []
        for _ in range(runs_per_size):
            arr = [random.randint(0, 1_000_000) for _ in range(n)]

            start = time.perf_counter()
            heap_sort(arr)
            end = time.perf_counter()

            tempos.append(end - start)

        media = statistics.mean(tempos)
        desvio = statistics.stdev(tempos)  

        resultados.append({
            'tamanho': n,
            'tempos': tempos,
            'media': media,
            'desvio': desvio
        })

    return resultados


if __name__ == "__main__":
    tamanhos = [100, 500, 1000, 5000, 10000, 50000]
    resultados = run_experiment(tamanhos, runs_per_size=20)

    print("n\tmedia (s)\tdesvio-padrão (s)")
    for r in resultados:
        print(f"{r['tamanho']}\t{r['media']:.6f}\t{r['desvio']:.6f}")
